Add revocation option to the generated CRED_DEF section (#32)
AnonCreds-WG/anoncreds-spec: spec/data_flow_setup.md

    end
```

#### SCHEMA Publisher: Publish SCHEMA Object

Each type of AnonCred credential is based on a [[ref: SCHEMA]] published to a Verifiable

> Data Registry (VDR), an instance of Hyperledger Indy in this version of

Reviewer: I believe we are missing a definition in the terminology section for "Verifiable Data Registry (VDR)" or is ist something that we assume the reader already knows?

> AnonCreds. The [[ref: SCHEMA]] is defined and published by the [[ref: SCHEMA]] Publisher. Any issuer

Reviewer: ```suggestion
AnonCreds. The [[ref: SCHEMA]] is defined and published by the [[ref: SCHEMA publisher]]. Any issuer
```

> who can reference the [[ref: SCHEMA]] (including the [[ref: SCHEMA]] Publisher) MAY issue

Reviewer: ```suggestion
who can reference the [[ref: SCHEMA]] (including the [[ref: SCHEMA publisher]]) MAY issue
```

Author: Resolved -- checked the entire document and updated all instances in prose (e.g. excluding the diagram).

credentials of that type by creating and publishing a [[ref: CRED_DEF]] based on the
[[ref: SCHEMA]]. This part of the specification covers the operation to create and
publish a [[ref: SCHEMA]]. The flow of operations to publish a [[ref: SCHEMA]] is illustrated in
the `SCHEMA Publisher: Publish SCHEMA` section of the [AnonCreds Setup Data
Flow](#anoncreds-setup-data-flow) sequence diagram.

> The [[ref: SCHEMA]] is a JSON structure that can be manually constructed containing the

Reviewer 2: ```suggestion
The [[ref: SCHEMA]] is a JSON structure that can be manually constructed, containing the
```

#### Issuer Create and Publish CRED_DEF Object

Each Issuer of a credential type (e.g. one based on a specific [[ref: SCHEMA]]) needs to
create a [[ref: CRED_DEF]] for that credential type. The flow of operations to create and
publish a [[ref: CRED_DEF]] is illustrated in the `Issuer: Create, Store and Publish CRED_DEF`
section of the [AnonCreds Setup Data Flow](#anoncreds-setup-data-flow) sequence
diagram.

In AnonCreds, the [[ref: CRED_DEF]] and [[ref: CRED_DEF]] identifier include the following elements.

* A link to the Issuer of the credentials via the DID used to publish the
  [[ref: CRED_DEF]].
* A link to the [[ref: SCHEMA]] upon which the [[ref: CRED_DEF]] is based (the credential type).
* A set of public/private key pairs, one per attribute (claim) in the
  credential. The private keys will later be used to sign the claims when
  credentials to be issued are created.
* Other information necessary for the cryptographic signing of credentials.
* Information necessary for the revocation of credentials, if revocation is to
  be supported by the Issuer for this type of credential.

We'll initially cover the generation and data for a [[ref: CRED_DEF]] created without the

> option of revoking credentials. Following that, ([in this

Reviewer: ```suggestion
option of revoking credentials. Following that, [in this
section](#generating-a-cred_def-with-revocation-support),
```

> option of revoking credentials. Following that, ([in this
> section](#generating-a-cred_def-with-revocation-support)), we describe the

Reviewer: ```suggestion
option of revoking credentials. Following that, [in this
section](#generating-a-cred_def-with-revocation-support), we describe the
```

additions to the generation process and data structures when support for
revocations is activated for a given [[ref: CRED_DEF]].

##### Retrieving the SCHEMA Object

Prior to creating a [[ref: CRED_DEF]], the Issuer must get an instance of the
[[ref: SCHEMA]] upon which the [[ref: CRED_DEF]] will be created, including the
identifier for the [[ref: SCHEMA]] in the form of a Hyperledger Indy `TXN_ID`.

> If the Issuer is also the [[ref: SCHEMA]] Publisher, they will already have the

Reviewer: ```suggestion
If the Issuer is also the [[ref: SCHEMA publisher]], they will already have the
```

> [[ref: SCHEMA]] (with `TXN_ID`). If not the Issuer must request that information

Reviewer 2: ```suggestion
[[ref: SCHEMA]] (with `TXN_ID`). If not, the Issuer must request that information
```

Author: Fixed.

* `primary` is the data used for generating credentials.
* `n` is the ...

> * `r` is a list of the attributes in the credential, with an associated public key for each.

Reviewer: ```suggestion
* `r` is a list of attribute names which are going to be used in the credential, each with an associated public key.
```

I`d rather emphasize that those are the "attribute names" in contrast to the attribute (values) itselfes.

Author: Good catch. Fixed.

>   * `master_secret` is in all [[ref: CRED_DEF]]s and is for signing attribute that will be put into the credential from the Holder.

Author: Looks good. I adjusted the phrasing a little more.

* `ref` is the `TXN_ID` on the Hyperledger Indy ledger for the [[ref: SCHEMA]] from which the list of attributes is pulled.
* `signature_type` is always `CL` in this version of AnonCreds.
* `tag` is the `tag` value (a string) passed in by the Issuer to an AnonCred's [[ref: CRED_DEF]] create and store implementation.

The [[ref: SCHEMA]] `TXN_ID` and `tag` items are used to form a `cred_def_id` for the [[ref: CRED_DEF]]. The `CRED_DEF_id` is namespaced by the Issuer of the CreDef,
as follows: `<issuer DID>:<object type>:<signature_type>:<SCHEMA TXN_ID>:tag>`. The elements of the identifier, separated by `:`'s are:

* `issuer DID`: The DID of the Issuer, the issuer of the [[ref: CRED_DEF]].
* `object type`: The type of object. `3` is used for [[ref: CRED_DEF]]s.
* `signature_type`: The `signature_type` item from the [[ref: CRED_DEF]].
* `SCHEMA TXN_ID`: The `ref` item from the [[ref: CRED_DEF]]
* `tag`: The `tag` item from the [[ref: CRED_DEF]].

##### Generating a CRED_DEF With Revocation Support

The issuer enables the ability to revoke credentials produced from a [[ref: CRED_DEF]] by
passing to the [[ref: CRED_DEF]] generation process the flag `support_revocation` as

> `true`. When revocation is to be supported, additional data related to

Reviewer 2: When revocation is supported or  when revocation is enabled ?  I can see in other parts also mentioned "revocation enabled".  Maybe we should be consistent and either use supported or enabled.

Reviewer: I agree. Per default revocation is supported by Indy. IMO enabling or disabling revocation purely depends on setting the revocation flag when creating the CRED_DEF

Author: Updated all places in the section where I had used "supported" when it should have been "enabled".  I also replaced an instance of "activated" with "enabled" to be consistent.
